src/Error/TypeError.hs

```diff
@@ -21,6 +22,7 @@ data TypeError = Mismatch      {t1 :: Type,  t2 :: Type, expr :: Expr ()}
                | Dereff        {name :: String, typ :: Type}
                | Uninitialized {name :: String}
                | SigBadFeq     {name :: String, sigType :: Type, feq :: Equation ()}
+               | InputMismatch {x_type :: Xtype, val :: Val}
```

Comment:
`val :: Val` should be a type, not a value. The type checker should be used to get the actual type based on the user input.